src/main/java/org/folio/circulation/services/AllowedServicePointsService.java

```diff
@@ -32,7 +56,91 @@ private CompletableFuture<Result<Map<RequestType, Set<String>>>> getAllowedServi
 
     log.debug("getAllowedServicePointsForItem:: parameters: request={}", request);
 
-    return ofAsync(new EnumMap<>(RequestType.class));
+    return userRepository.getUser(request.getRequesterId())
+      .thenCompose(r -> r.after(user -> lookupRequestPolicy(request, user)))
+      .thenCompose(r -> r.after(this::extractAllowedServicePoints));
+  }
+
+  private CompletableFuture<Result<RequestPolicy>> lookupRequestPolicy(
+    AllowedServicePointsRequest request, User user) {
+
+    log.debug("lookupRequestPolicy:: parameters: request={}, user={}", request, user);
+
+    if (user == null) {
+      log.error("lookupRequestPolicy:: user is null");
+    }
+
+    return itemRepository.fetchById(request.getItemId())
+      .thenCompose(r -> r.after(item -> requestPolicyRepository.lookupRequestPolicy(item, user)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> extractAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("extractAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    var allowedServicePointsInPolicy = requestPolicy.getAllowedServicePoints();
+    if (allowedServicePointsInPolicy.isEmpty()) {
+      log.info("extractAllowedServicePoints:: allowedServicePointsInPolicy is empty");
+
+      return fetchAllowedServicePoints(requestPolicy);
+    }
+
+    Set<String> allowedServicePointsIdsSet = allowedServicePointsInPolicy.values().stream()
+      .flatMap(Collection::stream)
+      .map(UUID::toString)
+      .collect(Collectors.toSet());
+
+    return servicePointRepository.filterIdsByServicePointsAndPickupLocationExistence(
+      allowedServicePointsIdsSet)
+        .thenCompose(r -> r.after(filteredSpIds -> includeOnlyExistingServicePoints(
+          allowedServicePointsInPolicy, filteredSpIds, requestPolicy)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> fetchAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("fetchAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    List<RequestType> allowedTypes = Arrays.stream(RequestType.values())
+      .filter(requestPolicy::allowsType)
+      .toList();
+    if (!allowedTypes.isEmpty()) {
```

Comment:
can we check `if (allowedTypes.isEmpty())` first, do the return, and then continue without redundant nesting?